### PR TITLE
[ImportVerilog] Update to slang 4

### PR DESCRIPTION
**Pull request by albertethon**
The modifications added the necessary macro definitions for the slang runtime, while importVerilog was missing them.

### CODE REVIEW

**Comment by fabianschuiki** [comment]
Hey @albertethon, thanks a lot for working on this!

Does Slang 3.0 really _require_ Boost? Can it be disabled? Boost is a pretty heavyweight dependency. Would be great if we could avoid this somehow.

Could you tweak this PR such that it only contains the minimal changes necessary to add Slang 3.0 support, an none of the other expression support and the like? Thanks a lot :raised_hands: :pray:

**Comment by albertethon** [comment]
Thanks for your suggestion. The current change deletes some original macro definitions, but limits the version of slang to a 3.0.0, which can ensure a little stability.

File: lib/Conversion/ImportVerilog/CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,23 @@ else ()
   add_compile_options(-Wno-cast-qual)
   # visitor switch statements cover all cases but have default
   add_compile_options(-Wno-covered-switch-default)
+  # several macro flags slang used
+  add_compile_definitions(DEBUG)
+  add_compile_definitions(BOOST_ALL_NO_LIB)
+  add_compile_definitions(SLANG_STATIC_DEFINE)
+  add_compile_definitions(SLANG_USE_BOOST)
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/CMakeLists.txt]:
```suggestion
  # several macro flags slang used
  add_compile_definitions(DEBUG)
  add_compile_definitions(SLANG_STATIC_DEFINE)
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -540,7 +540,13 @@ if(CIRCT_SLANG_FRONTEND_ENABLED)
...
     set(CMAKE_CXX_FLAGS ${original_CXX_FLAGS})
   else()
-    find_package(slang 2.0 REQUIRED)
+    # If you use the external way, you should use configure and build slang 
+    # with 3.0.0 version. Clone the slang and checkout to v3.0
+    # if you failed to build the slang v3.0 due to
+    # conflicts about package version used in slang, you can run with
+    # -DFETCHCONTENT_TRY_FIND_PACKAGE_MODE=NEVER to force it to download the
+    # correct version and build with that.
+    find_package(slang 3.0.0 EXACT)
```
Comment (fabianschuiki) [inline CMakeLists.txt]:
```suggestion
    # If you use the external way, you should use configure and build slang 
    # with 4.0.0 version. Clone the slang and checkout to v4.0
    # if you failed to build the slang v4.0 due to
    # conflicts about package version used in slang, you can run with
    # -DFETCHCONTENT_TRY_FIND_PACKAGE_MODE=NEVER to force it to download the
    # correct version and build with that.
    find_package(slang 4.0.0 EXACT)
```

**Review by fabianschuiki — COMMENTED** [review]
Can you try to switch this to slang v4.0 such that we don't have to pull in the entirety of boost as a dependency? Thanks :smiley:

**Comment by albertethon** [comment]
Thanks a lot.
When I switch to slang v4.0, I found that some change happen.
e.g. There is no member named `buffers` in `slang::driver::Driver`
which is used in `circt/lib/Conversion/ImportVerilog/ImportVerilog.cpp:166:12`
```cpp
driver.buffers.push_back(slangBuffer);
```
Then build failed.
The [commit](https://github.com/MikePopoloski/slang/commit/22205e472a25882db262daa6e2910b6c8357a1e5) about this change is here:
https://github.com/MikePopoloski/slang/commit/22205e472a25882db262daa6e2910b6c8357a1e5

I'm not sure about how to change this code correctly in your project.
I would be very grateful if you could make this change or tell me how to do it.

**Comment by albertethon** [comment]
I tried to add 
`driver.sourceLoader.addFiles(mlirBuffer->getBufferIdentifier());`
to replace
`driver.buffers.push_back(slangBuffer); `
while using slang v4.0, and it works well !
If you have better choice, please tell me, thank you.

**Comment by fabianschuiki** [comment]
Thanks for tackling this! :tada:

File: lib/Conversion/ImportVerilog/ImportVerilog.cpp [hunk 3]
```diff
@@ -163,7 +163,9 @@ mlir::OwningOpRef<mlir::ModuleOp> circt::importVerilog(SourceMgr &sourceMgr,
     const llvm::MemoryBuffer *mlirBuffer = sourceMgr.getMemoryBuffer(i + 1);
     auto slangBuffer = driver.sourceManager.assignText(
         mlirBuffer->getBufferIdentifier(), mlirBuffer->getBuffer());
-    driver.buffers.push_back(slangBuffer);
+    // driver.buffers.push_back(slangBuffer);
```
Comment (fabianschuiki) [inline lib/Conversion/ImportVerilog/ImportVerilog.cpp]:
```suggestion
```

**Comment by fabianschuiki** [comment]
I'm trying a build in #21. If that works I'll land your PR via that branch 👍

**Comment by fabianschuiki** [comment]
Thanks for the great help to get v3.0 set up @albertethon! 🥳  Really appreciate it!

**Comment by albertethon** [comment]
I'll organize my other work later.  Thanks for your response !  Learned a lot :smile:

**Pull request closed without merging**